[EDR Workflows] Add Crowdstrike connector and Actions

tomsonpl opened the pull request:
This PR introduces new connector - Crowdstrike (technical preview), it is hidden behind a `crowdstrikeConnectorOn` feature flag in `stack_connectors`.

 
- [x] Functionality is hidden behind multiple featureFlags.
- [x] Add a connector accepting clientId and clientSecret
- [x]  Add subActions of HostActions (isolate and release) and getAgentDetails (commented out for now)
- [x] Add a way to authenticate with Crowdstrike to get access_token that has to be passed on to other requests

<img width="2995" alt="Zrzut ekranu 2024-04-17 o 17 48 52" src="https://github.com/elastic/kibana/assets/16632552/0835e416-dab4-4211-bfa9-81ddbdae0573">
<img width="2999" alt="Zrzut ekranu 2024-04-5 o 15 06 19" src="https://github.com/elastic/kibana/assets/16632552/f7d2d20b-d1d8-4be6-a4bb-2633974e14a5">

tomsonpl commented:
/ci

tomsonpl commented:
/ci

elasticmachine commented:
Pinging @elastic/security-defend-workflows (Team:Defend Workflows)

tomsonpl commented:
@elasticmachine merge upstream

paul-tavares commented on x-pack/plugins/actions/server/sub_action_framework/sub_action_connector.ts:
```diff
@@ -84,7 +84,7 @@ export abstract class SubActionConnector<Config, Secrets> {
 
   private getHeaders(headers?: AxiosRequestHeaders): Record<string, AxiosHeaderValue> {
-    return { ...headers, 'Content-Type': 'application/json' };
+    return { 'Content-Type': 'application/json', ...headers };
```
Was this intentional?

tomsonpl replied:
Yes, before, the `Content-Type': 'application/json'  couldn't be overwritten.

paul-tavares commented on x-pack/plugins/stack_connectors/common/crowdstrike/schema.ts:
```diff
@@ -0,0 +1,199 @@
+  clientSecret: schema.string(),
+});
+
+// TODO TC: Response Schemas - most is commented out, because we are not able to test it against a real endpoint just yet.
```
suggestion:

If the data you have commented out below is "real" data (meaning: it was defined based on the data you captured while you had access to the customer's environment), then I would actually uncomment it all - we need to have something to work with as we move forward to mock a crowdstrike API server.

For reponse data properties that we don't forsee needing or for those that you are not sure are always returned, I would say wrap them in `schema.nullable()`. That along with the options to allow unknowns should hopefully prevent the Sub-actions framework from emitting response schema validation failures.

paul-tavares replied:
should probably remove this `todo` now

paul-tavares commented on x-pack/plugins/stack_connectors/public/connector_types/crowdstrike/crowdstrike.ts:
```diff
@@ -0,0 +1,67 @@
+    selectMessage: i18n.translate(
+      'xpack.stackConnectors.security.crowdstrike.config.selectMessageText',
+      {
+        defaultMessage: 'Execute CrowdStrike scripts',
```
"scripts" sounds strange to me. Maybe set the message to "Execute CrowdStrike response actions" ??

paul-tavares commented on x-pack/plugins/stack_connectors/server/connector_types/crowdstrike/crowdstrike.ts:
```diff
@@ -0,0 +1,179 @@
+} from '../../../common/crowdstrike/schema';
+import { SUB_ACTION } from '../../../common/crowdstrike/constants';
+
+export const API_PATH = 'https://api.crowdstrike.com';
```
We chatted briefly offline, but wanted to provide the comment here for visibility.

Its strange to me that we are hardcoding the URL to the API. This (IMO) can have some drawbacks and limitations from a customer's standpoint, including:

- Support on-prem instances of CrowdStrike (if thats a thing)
- Being unable to proxy requests through a corporate gateway (ex. for highly secured environments)
- Supporting air-gaped environments

paul-tavares replied:
Anything on this comment?

my suggestion: add it to the UI (form) default it to this URL. That gives the user the opportunity to change it if in fact there could be a different URL for the API

paul-tavares commented on x-pack/plugins/stack_connectors/server/connector_types/crowdstrike/crowdstrike.ts:
```diff
@@ -0,0 +1,179 @@
+  CrowdstrikeConfig,
+  CrowdstrikeSecrets
+> {
+  private static token: string | null;
+  private static tokenExpiryTimeout: NodeJS.Timeout;
```
Suggestion: add some jsdocs here on these two properties explaining why we are doing this?

paul-tavares replied:
Forget to add some docs?

paul-tavares commented on x-pack/plugins/stack_connectors/server/connector_types/crowdstrike/crowdstrike.ts:
```diff
@@ -0,0 +1,179 @@
+        return this.crowdstrikeApiRequest(req, true);
+      }
+
+      throw new Error(error.message);
```
can we improve the Error response here (I been meaning to do the same for SentinelOne).
One thing that I have found to be very difficult to work with is the fact that the Response body of the failure is not accessible and that usually provide more information than just (for example): `Error: Status code: 400. Message: API Error: Bad Request`

IMO, it would be nice to have a Crowdstrike specific Error class (sub-class from `Error`) that could include additional data associated with the HTTP failure. I'm not suggesting the entire `Response` object be returned, as that would expose some data that we definitely  don't want to expose (API keys).

tomsonpl replied:
something like this #8e816f0d65fe6a99011bbd08e5ac20f3574e6c13 ?

paul-tavares commented on x-pack/plugins/stack_connectors/server/connector_types/crowdstrike/crowdstrike.ts:
```diff
@@ -0,0 +1,179 @@
+    if (error.response.status === 401) {
+      return 'Unauthorized API Error';
+    }
+    return `API Error: ${error.response?.statusText}`;
```
for all of these message: They are not very useful and it would be very difficult tl debug this in production. Is there any other helpful data we can include that can help determine the root cause?

paul-tavares replied:
FYI: I would still suggest adding `JSON.stringify(error.response.data ?? {})` to all of these messages. If the code reaches this far, then we still don['t have much to determine what the problem might have been.

paul-tavares reviewed (commented):
Thanks for working ont his. I left several comments through the code in this PR. Let me know what you think. I also checked out the PR and validated the UI changes and Authz via API - looks good.

ymao1 commented on x-pack/plugins/stack_connectors/server/connector_types/crowdstrike/crowdstrike.ts:
```diff
@@ -0,0 +1,191 @@
+    }
+
+    if (!error.response?.status) {
+      return 'Unknown API Error';
```
if there is no status, does that automatically mean there's no additional information? could we stringify the error object maybe to provide additional error context?

tomsonpl replied:
We should get an error in the check above, but yeah - I think there's not much to add here. I believe we can improve the errors with the next steps though.

ymao1 commented on x-pack/test/alerting_api_integration/security_and_spaces/group2/tests/actions/connector_types/crowdstrike.ts:
```diff
@@ -0,0 +1,233 @@
+
+            expect({ status, message, connector_id }).to.eql({
+              status: 'error',
+              message: 'an error occurred while running the action',
```
are there plans to add some sort of server simulator to simulate responses for these subactions?

tomsonpl replied:
Absolutely, as a next step there's another PR that will mock responses in Cypress. However we're investigating a way to have a generic fake server for all 3rd party bidirectional vendors. It might take a moment to get it done though.

tomsonpl commented on x-pack/plugins/stack_connectors/server/connector_types/crowdstrike/crowdstrike.ts:
```diff
@@ -0,0 +1,191 @@
+    });
+  }
+
+  private async getTokenRequest() {
```
Thank you, indeed it looks similar - definitely worth taking a closer look at it with the next iteration 👍

tomsonpl commented on x-pack/plugins/stack_connectors/common/crowdstrike/schema.ts:
```diff
@@ -0,0 +1,199 @@
+  command: schema.oneOf([schema.literal('contain'), schema.literal('lift_containment')]),
+  actionParameters: schema.maybe(schema.object({}, { unknowns: 'allow' })),
+  ids: schema.arrayOf(schema.string()),
+  alertIds: schema.maybe(schema.arrayOf(schema.string())),
```
I have trouble understanding. Are you saying that the subactions we invoke are stored in saved objects somewhere? 
I think we do not need that for our needs (@paul-tavares please correct me if I am wrong), so I will adjust the code to your preference. 
Should I then wrap all of them in schema.nullable() ?

ymao1 replied:
During normal rule execution, if a rule triggers an action, the params for that action are stored in an `action_task_params` saved object and then the action executor, whenever it runs, will read this saved object from persistence. 

The other way of executing actions is by directly calling `execute` via the `actionsClient` and this does not require this extra saved object layer. I'm guessing that the bidirectional connectors will always be executed this second way which doesn't require the saved object? If that's the case, the `schema.maybe` seems fine

paul-tavares replied:
Correct. for now, we are not plugged into the whole "rule" processing. We're using the connector directly via the `ActionClient` and `UnsecuredActionClient`

@ymao1 - thanks for that explanation. I had been wondering they I kept seeing so much use of `schema.nullable()` instead of `schema.maybe()`.

tomsonpl commented:
@ymao1 Thank you for the review, appreciate the help - I like the suggested `getOAuthClientCredentialsAccessToken` util a lot! 
I understand your concerns about server to mock actions, types not being strict etc. This is still an ongoing work, we are having hard time with getting real access to Crowdstrike (we were able to use it just for a few hours and our development is based on the knowledge we gathered then), so we are not planning to release this anytime soon. This PR is the first step to enable other work that has to be done and then tested. So I hope if this implementation doesn't break any of your crucial practices and you'd be fine with accepting it in a `not perfect` state. 

I hope the small changes I made addressed your concerns. Please tell me if there's something more I can do. Of course I am planning to add schema.nullable() if you say we should 👍

ymao1 commented on x-pack/plugins/actions/server/action_type_registry.ts:
```diff
@@ -230,8 +230,10 @@ export class ActionTypeRegistry {
-        .filter(([actionTypeId]) => (featureId ? actionTypeId !== '.sentinelone' : true))
+        // Temporarily don't return SentinelOne and Crowdstrike connector for Security Solution Rule Actions
+        .filter(([actionTypeId]) =>
+          featureId ? actionTypeId !== '.sentinelone' && actionTypeId !== '.crowdstrike' : true
```
can use `!isBidirectionalConnectorType` here :)

ymao1 commented on x-pack/plugins/actions/server/authorization/actions_authorization.ts:
```diff
@@ -77,9 +78,9 @@ export class ActionsAuthorization {
+            // SentinelOne and Crowdstrike sub-actions require that a user have `all` privilege to Actions and Connectors.
             // This is a temporary solution until a more robust RBAC approach can be implemented for sub-actions
-            actionTypeId === '.sentinelone'
+            actionTypeId === '.sentinelone' || actionTypeId === '.crowdstrike'
```
can use `isBidirectionalConnectorType` here :)

ymao1 reviewed (approved):
LGTM

tomsonpl commented:
Oh I see, thank you for explanation! 
We are investigating a way forward for enabling automated SentinelOne actions within a rule, but the first iteration will most probably be with the use of `Response Actions` and not `Actions`. 
For a long term plan we are aiming at possibly migrating to `System Actions` when they are ready and stable.

paul-tavares commented on x-pack/plugins/actions/server/lib/bidirectional_connectors.ts:
```diff
@@ -0,0 +1,15 @@
+ */
+
+const BIDIRECTIONAL_CONNECTOR_TYPES = ['.sentinelone', '.crowdstrike'];
+export const isBidirectionalConnectorType = (type: string | undefined) => {
```
Nice!

paul-tavares commented on x-pack/plugins/stack_connectors/common/crowdstrike/schema.ts:
```diff
@@ -0,0 +1,205 @@
+            platform_name: schema.maybe(schema.string()),
+            policies: schema.maybe(
+              schema.arrayOf(
+                schema.object({
```
Suggestion: add `{ unknowns: 'allow' }` to this `schema.object()` definition 

Same comment for others definitions below. Adding this will just ensure that if a response from the 3rd party system returns a new value that we are not aware of it will not cause the sub-actions framework to throw an error.

paul-tavares replied:
did you miss this `schema.object()`? I see the `{ unknowns: 'allow' }` in other places below, but not here.

paul-tavares commented on x-pack/plugins/stack_connectors/common/crowdstrike/schema.ts:
```diff
@@ -0,0 +1,205 @@
+    access_token: schema.string(),
+    expires_in: schema.number(),
+    token_type: schema.string(),
+    // fields also existing according to the docs
+    // id_token: schema.maybe(schema.string()),
+    // issued_token_type: schema.maybe(schema.string()),
+    // refresh_token: schema.maybe(schema.string()),
+    // scope: schema.maybe(schema.string()),
```
did you forget to remove or uncomment these?

paul-tavares commented on x-pack/plugins/stack_connectors/server/connector_types/crowdstrike/crowdstrike.test.ts:
```diff
@@ -0,0 +1,214 @@
+
+  beforeEach(() => {
+    // @ts-expect-error private static - but I still want to reset it
+    CrowdstrikeConnector.token = null;
```
This is interesting (surprising too).  I guess the transpiler for TS to JS does not define the property as "private" class member, thus why you are able to do this

tomsonpl replied:
🤷  :)

paul-tavares commented on x-pack/plugins/stack_connectors/server/connector_types/crowdstrike/crowdstrike.test.ts:
```diff
@@ -0,0 +1,214 @@
+  });
+
+  describe('executeHostActions', () => {
+    afterEach(() => {
+      jest.clearAllMocks();
+    });
```
FYI: redundant. Its already done by the top-level `afterEach()`

paul-tavares commented on x-pack/test/alerting_api_integration/security_and_spaces/group2/tests/actions/connector_types/crowdstrike.ts:
```diff
@@ -0,0 +1,233 @@
+
+      let connectorId: string;
+
+      const createUser = async ({
```
hmm. this code looks familiar 😉 😄 

at some point, it may be time to create an FTR service for this stuff and then re-use it across SentinelOne and Crowstrike tests

tomsonpl commented:
Thanks @paul-tavares , very valuable suggestions 👍 I applied all of them here: [3fd0b85](https://github.com/elastic/kibana/pull/180175/commits/3fd0b8561d47659d6b361f6868aeaee06228323f) 
@ymao1 could you also take a look at the mentioned commit? I added a `url` field into the connectors config. 

Thank you both!

paul-tavares reviewed (approved):
👍

ymao1 commented:
> @ymao1 could you also take a look at the mentioned commit? I added a url field into the connectors config.

URL changes look good! Verified that it is validated on the UI level and at the API level and respects the allowedHosts config
